src/main/java/ch/njol/skript/entity/DroppedItemData.java

```diff
@@ -111,36 +127,68 @@ public EntityData getSuperType() {
 	}
 	
 	@Override
-	public String toString(final int flags) {
-		final ItemType[] types = this.types;
+	public String toString(int flags) {
 		if (types == null)
 			return super.toString(flags);
-		final StringBuilder b = new StringBuilder();
-		b.append(Noun.getArticleWithSpace(types[0].getTypes().get(0).getGender(), flags));
-		b.append(m_adjective.toString(types[0].getTypes().get(0).getGender(), flags));
-		b.append(" ");
-		b.append(Classes.toString(types, flags & Language.NO_ARTICLE_MASK, false));
-		return "" + b.toString();
+		int gender = types[0].getTypes().get(0).getGender();
+		return Noun.getArticleWithSpace(gender, flags) +
+				m_adjective.toString(gender, flags) +
+				" " +
+				Classes.toString(types, flags & Language.NO_ARTICLE_MASK, false);
 	}
-	
-//		return ItemType.serialize(types);
+
 	@Override
 	@Deprecated
-	protected boolean deserialize(final String s) {
+	protected boolean deserialize(String s) {
 		throw new UnsupportedOperationException("old serialization is no longer supported");
-//		if (s.isEmpty())
-//			return true;
-//		types = ItemType.deserialize(s);
-//		return types != null;
 	}
 	
 	@Override
-	protected boolean equals_i(final EntityData<?> obj) {
-		if (!(obj instanceof DroppedItemData))
+	protected boolean equals_i(EntityData<?> otherData) {
+		if (!(otherData instanceof DroppedItemData))
 			return false;
-		return Arrays.equals(types, ((DroppedItemData) obj).types);
+		return Arrays.equals(types, ((DroppedItemData) otherData).types);
 	}
-	
+
+	@Override
+	public boolean canSpawn(@Nullable World world) {
+		return (HAS_JAVA_CONSUMER_DROP || HAS_BUKKIT_CONSUMER_DROP) && types != null && types.length > 0 && world != null;
+	}
+
+	@Override
+	@SuppressWarnings({"deprecation"})
+	public @Nullable Item spawn(Location location, @Nullable Consumer<Item> consumer) {
+		World world = location.getWorld();
+		if (!canSpawn(world))
+			return null;
+		assert types != null && types.length > 0;
+
+		final ItemType t = CollectionUtils.getRandom(types);
+		assert t != null;
+		ItemStack stack = t.getItem().getRandom();
+
+		Item item;
+		if (consumer == null) {
+			item = world.dropItem(location, stack);
+		} else if (HAS_JAVA_CONSUMER_DROP) {
+			item = world.dropItem(location, stack, consumer);
+		} else if (HAS_BUKKIT_CONSUMER_DROP) {
+			try {
+				assert BUKKIT_CONSUMER_DROP != null;
+				item = (Item) BUKKIT_CONSUMER_DROP.invoke(world, location, stack, (org.bukkit.util.Consumer<Item>) consumer::accept);
+			} catch (InvocationTargetException | IllegalAccessException e) {
+				if (Skript.testing())
+					Skript.exception(e, "Can't spawn " + this.getName());
+				return null;
+			}
+		} else {
+			return null;
```

Comment:
can't we drop and run the consumer after?

Reply:
I'd have to check to make sure the behavior is the same, but I think that'd be ok